### PR TITLE
implement simple batching for paxos requests

### PR DESCRIPTION
**Pull request by rjullman**
this edits the old broken batching system that allowed
people to join in-flight batches after they had checked (or partially
checked) for leadership

### CODE REVIEW

File: leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionService.java [hunk 1]
```diff
@@ -366,6 +369,52 @@ public StillLeadingCall() {
...
+            int val = requestCount.incrementAndGet();
+            return isRequestCountValid(val);
+        }
+
+        public void awaitPopulation() {
+            try {
+                populationLatch.await();
+            } catch (InterruptedException e) {
+                log.error("Interrupted while waiting for a batch!", e);
+                Throwables.throwUncheckedException(e);
```
Comment (rjullman) [inline leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionService.java]: Should it fail here or just log an error and ignore the interrupt?

Comment (carrino) [inline leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionService.java]: i'm good with "throw Throwables.throwUncheckedException(e);"

File: leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionService.java [hunk 2]
```diff
@@ -366,6 +369,52 @@ public StillLeadingCall() {
...
+
+        public void fail() {
+            failed = true;
+        }
+        public void becomeReadable() {
+            populationLatch.countDown();
+        }
+        // End creator-only threads.
+
+        public boolean isFailed() {
```
Comment (carrino) [inline leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionService.java]: add comment // This must only be called after awaitPopulation().

**Comment by carrino** [comment]
looks good

**Comment by carrino** [comment]
looks good

**Pull request merged**